src/viewer.ts

```diff
@@ -1,248 +1,219 @@
+  ): void {
+    const raycastEvents = [
+      "click",
+      "dblclick",
+      "mousedown",
+      "mouseup",
+      "touchstart",
+      "touchend",
+    ] as const;
+    const eventNameLower: string = eventName.toLowerCase();
+
+    if (!(raycastEvents as unknown as string[]).includes(eventNameLower)) {
+      this.elemViewer.addEventListener(eventNameLower, callable, false);
+      return;
     }
 
-    public setModel(
-        meshUpdated?: SurfaceMesh,
-        mapUpdated?: MeshColors,
-    ) {
-        if (meshUpdated) {
-            this.surface.mesh = meshUpdated;
+    this.elemViewer.addEventListener(
+      eventNameLower as (typeof raycastEvents)[number],
+      (event: MouseEvent | TouchEvent) => {
+        const rect = this.elemViewer.getBoundingClientRect();
+        const mice = getClicks(event, rect);
+
+        const intersects = [];
+        for (let i = 0; i < mice.length; i++) {
+          this.raycaster.setFromCamera(mice[i], this.camera);
+          const inter = this.raycaster.intersectObjects(this.getModels());
+          intersects.push(inter[0]);
         }
-
-        if (mapUpdated) {
-            this.surface.colors = mapUpdated;
-        }
-
-        //const geometry = new THREE.SphereGeometry();
-        const geometry = new THREE.BufferGeometry();
-
-        const verts = this.surface.mesh.vertices;
-
-        let material: THREE.MeshLambertMaterial;
-        if (this.surface.colors.colors) {
-
-            // cols = [r,g,b,r,g,b,...]
-
-            geometry.setAttribute("color", new THREE.BufferAttribute(this.surface.colors.colors, 3));
-
-            material = new THREE.MeshLambertMaterial({
-                vertexColors: true,
-            });
-        } else {
-
-            material = new THREE.MeshLambertMaterial({
-                color: /*new THREE.Color(*/"orange"/*)*/,
-                vertexColors: false,
-            });
-
-            this.legend.remove();
-        }
-
-        geometry.setAttribute("position", new THREE.BufferAttribute(verts, 3));
-        geometry.setIndex(new THREE.BufferAttribute(this.surface.mesh.faces, 1));
-
-        geometry.computeVertexNormals();
-
-        const obj = new THREE.Mesh(geometry, material);
-        obj.rotation.x = -Math.PI / 2;
-        this.scene.add(obj);
-    }
-
-    public dispose(): void {
-        window.removeEventListener("resize", this.onWindowResize);
-    }
+        callable(event, intersects);
+      },
+    );
+  }
+
+  public addModel(surface: Surface): void {
+    const obj = surfaceToMesh(surface);
+    this.scene.add(obj);
+  }
+
+  public getModels(): THREE.Mesh[] {
+    return this.scene.children.filter(
+      (child) => child.type === "Mesh",
+    ) as THREE.Mesh[];
+  }
+
+  public deleteModel(surface: Surface): void {
+    const models = this.getModels();
+    const surface_mesh = surfaceToMesh(surface);
+
+    models.map((model) => {
+      if (isMeshEqual(model, surface_mesh)) {
+        this.scene.remove(model);
+      }
```

Comment:
I see, this is why you added it. I've made the addModel() method above return the THREE.Mesh, this way we dont need this at all. (At the expense of the user needing to keep track of the mesh reference, which I think is fair)

Reply:
Good point, will implement this.

Reply:
This function now just uses the object reference as input and is essentially a convenience function for this.scene.remove